pkg/server/registry/images/strategy.go

```diff
@@ -3,177 +3,156 @@ package images
-	"github.com/acorn-io/acorn/pkg/remoteopts"
+	"github.com/acorn-io/acorn/pkg/client"
 	"github.com/acorn-io/acorn/pkg/tables"
 	tags2 "github.com/acorn-io/acorn/pkg/tags"
+	"github.com/acorn-io/mink/pkg/strategy"
+	"github.com/acorn-io/mink/pkg/strategy/remote"
+	"github.com/acorn-io/mink/pkg/strategy/translation"
 	"github.com/acorn-io/mink/pkg/types"
 	"github.com/google/go-containerregistry/pkg/name"
-	"github.com/google/go-containerregistry/pkg/v1/remote"
-	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
-	corev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
 	"k8s.io/apiserver/pkg/registry/rest"
-	"k8s.io/apiserver/pkg/storage"
 	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
 )
 
 type Strategy struct {
+	strategy.CompleteStrategy
 	rest.TableConvertor
-	client kclient.WithWatch
-}
-
-func NewStrategy(c kclient.WithWatch) *Strategy {
-	return &Strategy{
-		TableConvertor: tables.ImageConverter,
-		client:         c,
-	}
-}
 
-func (s *Strategy) Get(ctx context.Context, namespace, name string) (types.Object, error) {
-	return s.ImageGet(ctx, namespace, name)
+	client        kclient.Client
+	clientFactory *client.Factory
 }
 
-func (s *Strategy) List(ctx context.Context, namespace string, opts storage.ListOptions) (types.ObjectList, error) {
-	images, err := s.ImageList(ctx, namespace)
+func NewStrategy(c kclient.WithWatch, clientFactory *client.Factory) (strategy.CompleteStrategy, error) {
+	storageStrategy, err := newStorageStrategy(c)
 	if err != nil {
 		return nil, err
 	}
-	return &images, nil
+	return NewStrategyWithStorage(c, clientFactory, storageStrategy), nil
 }
 
-func (s *Strategy) New() types.Object {
-	return &apiv1.Image{}
+func NewStrategyWithStorage(c kclient.WithWatch, clientFactory *client.Factory, storageStrategy strategy.CompleteStrategy) strategy.CompleteStrategy {
+	return &Strategy{
+		TableConvertor:   tables.ImageConverter,
+		CompleteStrategy: storageStrategy,
+		client:           c,
+		clientFactory:    clientFactory,
+	}
 }
 
-func (s *Strategy) NewList() types.ObjectList {
-	return &apiv1.ImageList{}
+func newStorageStrategy(kclient kclient.WithWatch) (strategy.CompleteStrategy, error) {
+	return translation.NewTranslationStrategy(
+		&Translator{},
+		remote.NewRemote(&v1.ImageInstance{}, &v1.ImageInstanceList{}, kclient)), nil
 }
 
-func (s *Strategy) ImageList(ctx context.Context, namespace string) (apiv1.ImageList, error) {
-	if namespace != "" {
-		return s.forNamespace(ctx, namespace)
-	}
-
-	var (
-		result     apiv1.ImageList
-		namespaces = &corev1.NamespaceList{}
-	)
-
-	err := s.client.List(ctx, namespaces)
-	if err != nil {
-		return result, err
-	}
+func (s *Strategy) Validate(ctx context.Context, obj runtime.Object) (result field.ErrorList) {
```

Comment:
After darren's refactor, this Validate method should completely go away, since users wont be allowed to Create images, just ImageInstances should be created by the backend. Validate maps to Create. ValidateUpdate maps to update.

You'll have to migrate the logic to validateUpdate

Reply:
added a TODO